### DISCUSSION TITLE
How do I set the locale if I don't have a param?

### DISCUSSION THREAD

**Discussion by tkhobbes** [ddnexus/pagy#526] [Q&A] [answered]
For now I am setting a default locale (de) in my application.rb, as I am not yet sure how I will implement locales and locale switching in my app.
This works as expected everywhere in my app, i. e. the de locales are used everywhere.
I am loading two locales for pagy, via config/initializers/pagy.rb:

`Pagy::I18n.load({ locale: 'en' },
                { locale: 'de' })`

When I use a URL like localhost:3000?locale=de and the following code, pagy i18n behaves as expected:
`before_action { @pagy_locale = params[:locale] }`

But now I want to do something like this - however that is not working (pagy shows en, which is the first loaded locale, i. e. it doesn't recognize that it should use de):
`before_action { @pagy_locale = params[:locale] || I18n.default_locale }`

When I log @pagy_locale, it correctly seems to be set to "de" - why isn't pagy picking that up?

What am I missing?

**Comment by ddnexus** [comment]
Not sure I get your problem. Could you explain it explicitly?

If I have to guess... maybe the following sentence from the doc may be considered:

If the `@pagy_locale` is nil or missing pagy will serve the first locale you set in the configuration.

**Reply by tkhobbes** [reply]
Thanks for the quick reply - this is my issue: If I do something like `<%= @pagy_locale %>` somewhere on a view, it says 'de'. So it seems to be set. BUT pagy doesn't pick the locale up - the buttons are labelled in english, not in german.
When I explicitely use http://localhost:3000?locale=de, @pagy_locale is still "de" (obviously), AND pagy shows the german labels.

I do not want to set the locale explicitely in the params for now. I have set it in my application.rb as default_locale to "de".
Whether or not I do anything in a before_action lambda... pagy always shows the english labels on the buttons though.

So my question is: Why does @pagy_locale correctly show "de" but pagy itself doesn't pick up the locale for displaying the buttons?

Hope that is more clear.

**Comment by ddnexus** [comment]
So you are saying that it works only when there is a locale param in the request. 

Maybe.

I will be able to check that detail in a few days.

Meanwhile, if you just invert the order of the locales hash arguments in the initializer, it should use `de` (first locale argument) when there is no locale param, so at least it will work the way you want. (and without using `|| I18n.default_locale`)

**Reply by tkhobbes** [reply]
Thanks - yes confirmed, my observation is exactly this.

**Comment by ddnexus** [comment, marked as answer]
OK, I got a few problems recently. Sorry for the late answer.

The easiest and recommended way to work with default `de` locale is setting `de` as the first locale:

```rb
Pagy::I18n.load({ locale: 'de' }, { locale: 'en' })
```

Then you can simply do:

```rb
before_action { @pagy_locale = params[:locale] }
```

There will be no `:locale` param for `de`, only for `en` (and/or any other locale that is not the first default).

> before_action { @pagy_locale = params[:locale] || I18n.default_locale }
> When I log @pagy_locale, it correctly seems to be set to "de" - why isn't pagy picking that up?

It could be a number of things: I should have a minimal example of your case to be able to understand what is going on in your code.

**Comment by tkhobbes** [comment]
Thank you. I will implement as per the above. If you want to check out my repository - it's here: [https://github.com/tkhobbes/ltb-rails](https://github.com/tkhobbes/ltb-rails) - I don't think I am doing anything special but you might figure out what is going on. Pagination is eg happening on books/index.html.erb.

**Reply by ddnexus** [reply]
Does it work with the recommended way, and does it fit your needs?

**Reply by tkhobbes** [reply]
It does work and it fits my needs. Still weird that the other approach did not work. But ok for the current use case.
Thanks for looking into it!

**Reply by ddnexus** [reply]
Glad that it works.

About the failure of the other approach: a possible suspect might be the `I18n.default_locale` not set as you would expect at the time of the `before_action`.
